src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -166,18 +168,20 @@ class Hyperspace(spark: SparkSession) {
 }
 
 object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  private[hyperspace] lazy val contexts =
+    mutable.WeakHashMap[SparkSession, ThreadLocal[HyperspaceContext]]()
```

Comment:
If the key is `SparkSession` and you are synchronizing on `contexts`, do you still need the value to be `ThreadLocal`?

Is this to support the scenario where two different threads share the same spark session? But even in this case, it seems that it will keep resetting if two threads call `getContext` back to back?

Reply:
> Is this to support the scenario where two different threads share the same spark session? 

I'm not sure about thread safety. One thread per session limitation might be good to have.

> But even in this case, it seems that it will keep resetting if two threads call getContext back to back?

I thought thread local variables stored in each thread context, so it would be ok..

Reply:
> I thought thread local variables stored in each thread context, so it would be ok..

But you are sharing a key in a map that's not thread local?

Reply:
> One thread per session limitation might be good to have.

We need to store more info then?

Reply:
Then how about (Thread.currentThread().getId, SparkSession) as a key?

Reply:
I think we can't "enforce" if the thread id is in the key. We can put it in the value `(Thread.currentThread().getId, HyperspaceContext)`, and throw an exception if current thread id is mismatching (and do not need to worry about `ThreadLocal` 😄 ).

Also, we need to limit the map size to say 16? To prevent the scenario where the user application can keep calling `newSession`. We can just throw an exception if it goes over the limit. WDYT?

Reply:
Yea - number of sessions limitation might be required but it can be configurable. 
And we could GC in Hyperspace (for every 5 mins?) - by ` spark.sparkContext.isStopped` and remove it. WDYT?

Reply:
I think we can just go simple and throw an exception if it goes over the limit (and the exception message can suggest to increase the limit if necessary).